### PR TITLE
feat(builder): collapse unconnected object sub-outputs

### PR DESCRIPTION
**Pull request by Dongyuan-Lithium**
The Builder nodes often expose object outputs as many sub-outputs (e.g., `Main Result.Text`, `Main Result.Html`, …). When those sub-outputs are unconnected, they take up significant canvas space and reduce usability. This PR collapses unconnected object sub-outputs by default while keeping connected outputs visible and allowing users to expand/collapse as needed.

### Changes 🏗️
- Collapse unconnected, non-top-level sub-outputs of object outputs by default in the Builder **FlowEditor** node output list.
- Add an expand/collapse chevron on object outputs plus a hidden-child count when collapsed.
- Keep connected sub-outputs visible even when the parent object output is collapsed.
- Persist collapse state per `nodeId + outputHandleId` in the FlowEditor store.
- Add unit tests covering output visibility/collapse behavior.

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - [x] `pnpm lint`
  - [x] `pnpm test:unit`
  - [x] Manual UI verification:
    - [x] Open `/build` (FlowEditor) and add **Execute Code** block
    - [x] Confirm `Main Result (object)` is collapsed by default and shows a chevron + hidden count
    - [x] Expand `Main Result` and verify sub-outputs appear
    - [x] Connect one `Main Result.*` sub-output, collapse parent, verify connected output remains visible

#### For configuration changes:
- [x] `.env.default` is updated or already compatible with my changes
- [x] `docker-compose.yml` is updated or already compatible with my changes
- [x] I have included a list of my configuration changes in the PR description (under **Changes**)

Configuration changes:
- None (frontend-only UI behavior change)

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds collapse/expand behavior for unconnected object sub-outputs in the FlowEditor, reducing visual clutter on the canvas. It also adds env var overrides for the `NEW_FLOW_EDITOR` and `BUILDER_VIEW_SWITCH` feature flags.

- Introduces `output-visibility.ts` — a well-structured pure-function module that builds a visibility tree from output schemas, accounting for collapse state and connected edges
- Refactors `OutputHandler.tsx` to use the new tree-based rendering with expand/collapse chevrons and hidden-child counts
- Adds `nodeOutputCollapsedStates` to the Zustand `nodeStore` with `set`, `toggle`, and `get` actions (defaults to collapsed)
- Adds unit tests (`output-visibility.test.tsx`) covering core visibility scenarios
- Modifies feature flag defaults to allow env var overrides for `NEW_FLOW_EDITOR` and `BUILDER_VIEW_SWITCH`

The logic separation into `output-visibility.ts` is clean and testable. The main areas to watch are the edge store subscription granularity (subscribes to all edges rather than node-specific ones) and the feature flag default changes which also affect Playwright mock mode.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

- This PR is safe to merge — the core logic is well-tested and the changes are frontend-only with no data model impact.
- Score of 4 reflects well-structured code with clean logic separation and unit tests covering key scenarios. Minor deductions for: subscribing to the full edges array (performance consideration, though consistent with prior behavior), redundant collapsed-state lookup logic, and subtle feature flag behavior change in Playwright mock mode. None of these are blockers.
- Pay attention to `use-get-flag.ts` — the default flag changes also affect Playwright mock mode behavior if the new env vars are set in CI.
</details>

<details><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[OutputHandler Component] -->|reads| B[edgeStore.edges]
    A -->|reads| C[nodeStore.nodeOutputCollapsedStates]
    A -->|calls| D[buildVisibleOutputTree]
    D -->|step 1| E[buildOutputTree - Parse schema into tree]
    D -->|step 2| F[buildVisibleNode - Apply collapse and connection logic]
    E -->|produces| G[OutputTreeNode list]
    G -->|input to| F
    F -->|produces| H[VisibleOutputTreeNode list]
    H -->|rendered by| I[renderOutputNode]
    I -->|if object with children| J[Show chevron toggle]
    I -->|if collapsed with hidden| K[Show hidden count badge]
    I -->|recursively render| L[Visible children]
    J -->|onClick| M[nodeStore.toggleOutputCollapsed]
    M -->|updates| C
```
</details>

<sub>Last reviewed commit: f5aba01</sub>

<sub>(2/5) Greptile learns from your feedback when you react with thumbs up/down!</sub>

<details><summary><h4>Context used (3)</h4></summary>

- Context from `dashboard` - autogpt_platform/CLAUDE.md ([source](https://app.greptile.com/review/custom-context?memory=6e9dc5dc-8942-47df-8677-e60062ec8c3a))
- Context from `dashboard` - autogpt_platform/frontend/CLAUDE.md ([source](https://app.greptile.com/review/custom-context?memory=39861924-d320-41ba-a1a7-a8bff44f780a))
- Context from `dashboard` - autogpt_platform/frontend/src/app/(platform)/build/components/FlowEditor/ARCHITECTURE_FLOW_EDITOR.md ([source](https://app.greptile.com/review/custom-context?memory=0c5511fe-9aeb-4cf1-bbe9-798f2093b748))
</details>

### CODE REVIEW

**Comment by Otto-AGPT** [comment]
Hey @Dongyuan-Lithium 👋

A few things before we can move forward:

1. **CLA**: Please sign the [Contributor License Agreement](https://cla-assistant.io/Significant-Gravitas/AutoGPT?pullRequest=12151) — it's required for all contributors.

2. **Merge conflicts**: This PR has conflicts with the `dev` branch. Please rebase or merge `dev` into your branch to resolve them.

3. **Review comments**: The greptile bot raised two valid points on `OutputHandler.tsx`:
   - The edges subscription could cause unnecessary re-renders for all nodes — consider filtering to only this node's edges in the selector
   - The `isOutputCollapsed` function has a redundant dual-source lookup that can be simplified

Looking forward to your updates! 🚀

**Review by autogpt-reviewer — COMMENTED** [review]
**PR #12151 — feat(builder): collapse unconnected object sub-outputs**
Author: Dongyuan-Lithium | Files: 5 (+422/-83) | Frontend-only

**🎯 Verdict: REQUEST_CHANGES**

### What This PR Does
Collapses unconnected object sub-outputs in the FlowEditor builder by default, reducing canvas clutter. Connected outputs remain visible when their parent is collapsed. Users can expand/collapse via a chevron toggle with a hidden-output count indicator. A new pure-function module (`output-visibility.ts`) handles the tree construction and visibility logic, with dedicated unit tests.

### Specialist Findings

🛡️ **Security** ✅ — No security concerns. React auto-escapes all rendered content, no `dangerouslySetInnerHTML`, no backend changes. Feature flag env overrides are build-time constants only. Store state is boolean-only, client-side. Clean.

🏗️ **Architecture** ✅ — Clean separation of visibility logic into a pure-function module with tests. Zustand store additions follow existing `nodeAdvancedStates` pattern. Two should-fix items: edge subscription subscribes to ALL edges (performance regression), and redundant `isOutputCollapsed` function duplicates store logic.

⚡ **Performance** ⚠️ — **Meaningful regression identified.** `useEdgeStore((s) => s.edges)` subscribes to the entire edge array — every edge change anywhere in the graph re-renders every `OutputHandler` on every node (O(N×E)). Additionally, `nodeOutputCollapsedStates` is selected globally (any node's collapse toggle re-renders all nodes), `buildVisibleOutputTree` is not memoized (recursive tree build on every frequent re-render), and `new Set()` is constructed every render. All fixable without architectural changes.

🧪 **Testing** ⚠️ — 5 unit tests cover core visibility happy paths but leave gaps: (1) no tests for 3 new nodeStore methods (`toggleOutputCollapsed` default behavior is non-obvious), (2) nested visibility assertion gap (expanding parent doesn't assert nested children stay hidden), (3) missing edge cases (empty properties, schema without `type: "object"`). E2E failures are infra-related (marketplace data seeding), not PR-related.

📖 **Quality** ⚠️ — Inline styles instead of Tailwind (`style={{ marginLeft: 6, cursor: "pointer" }}`), redundant `isOutputCollapsed` function duplicating store logic, `buildOutputTree` exported but only used internally, zero docstrings on public API, duplicate `getSchemaProperties` helper (already exists in `useSubAgentUpdate/helpers.ts`). Core logic is well-structured.

📦 **Product** ✅ — Solves a real usability pain point. Collapse-by-default is the right UX call. Connected outputs staying visible is excellent. Chevron + hidden count indicator provides adequate discoverability. Accessibility handled (aria-label, aria-expanded). State not persisted across page loads — acceptable since default (collapsed) is the desired state.

📬 **Discussion** ⚠️ — **PR is stalled.** CLA not signed (hard merge blocker). Merge conflicts with `dev` branch unresolved (9 days). 8 bot review comments (Greptile + CodeRabbit) — 0/8 addressed. Zero author responses in 9 days since initial push. Zero human maintainer reviews. Otto-AGPT nudged author on Feb 26 — no response.

🔎 **QA** ⏱️ — Browser automation session crashed during testing (technical failure). Unable to complete live validation. Environment was running correctly (Frontend :3200, Backend :8206, Storybook :6206 all healthy).

---

### 🔴 Blockers

1. **CLA not signed** — Hard merge blocker. Required for all contributors. Flagged Feb 18, unresolved.
2. **Merge conflicts with `dev` branch** — PR is in DIRTY merge state. Cannot merge.
3. **Edge subscription performance regression** (`OutputHandler.tsx:34`) — `useEdgeStore((s) => s.edges)` causes O(N×E) re-renders. Every edge change re-renders every OutputHandler on every node. Must use a filtered selector (e.g., `shallow` comparison on node-specific source handles).

### 🟠 Should Fix

1. **Global `nodeOutputCollapsedStates` selector** (`OutputHandler.tsx:39-41`) — Subscribes to collapse state for ALL nodes. Any toggle re-renders all OutputHandlers. Scope to `state.nodeOutputCollapsedStates[nodeId]`.
2. **Memoize `buildVisibleOutputTree`** (`OutputHandler.tsx:80-84`) — Recursive tree build on every render. Wrap in useMemo with proper deps (or extract to a selector).
3. **Redundant `isOutputCollapsed` function** (`OutputHandler.tsx:62-71`) — Duplicates `getOutputCollapsed` store method. Simplify to `nodeOutputCollapsedStates[nodeId]?.[handleId] ?? true`.
4. **Inline styles → Tailwind** (`OutputHandler.tsx:103`) — `style={{ marginLeft: 6, cursor: "pointer" }}` → `className="ml-1.5 cursor-pointer"`.
5. **Unexport `buildOutputTree`** (`output-visibility.ts:41`) — Internal helper, not used outside module.
6. **Add nodeStore method tests** — `toggleOutputCollapsed` default behavior (`?? true` → first toggle = expand) needs test coverage.
7. **Add nested visibility assertion** (`output-visibility.test.tsx:97`) — Expanding parent should assert nested children stay hidden when their immediate parent defaults to collapsed.
8. **Duplicate `getSchemaProperties`** — Consolidate with existing helper in `useSubAgentUpdate/helpers.ts`.

### 🟢 Nice to Have

1. Add docstrings to `buildVisibleOutputTree` and `VisibleOutputTreeNode` public API
2. Rename test file from `.tsx` to `.ts` (contains no JSX)
3. "Expand All / Collapse All" affordance for power users (future enhancement)
4. More descriptive hidden count label (e.g., "+3 outputs" instead of "(3)")

### Risk Assessment
Merge risk: **LOW** (frontend-only, behind feature flag) | Rollback: **EASY** (revert single commit)

### CI Status
- ✅ lint, types, integration_test, CodeQL, Snyk — all pass
- ❌ CLA pending, e2e_test fail (infra, not PR-related), Check PR Status fail, Vercel auth fail

### Process Note
Author has been unresponsive for 9 days. Zero commits since initial push (Feb 18). Zero responses to 8 bot review comments. If author remains unresponsive, this PR is a candidate for closure or adoption by a maintainer.

@ntindle Well-structured UX improvement that solves real canvas clutter — but needs CLA signed, conflicts resolved, edge subscription perf fix, and author engagement before it can merge.

File: autogpt_platform/frontend/src/app/(platform)/build/components/FlowEditor/nodes/OutputHandler.tsx [hunk 1]
```diff
@@ -26,94 +31,143 @@ export const OutputHandler = ({
...
-                        aria-label="info"
-                        tabIndex={0}
-                      >
-                        <InfoIcon />
-                      </span>
-                    </TooltipTrigger>
-                    <TooltipContent>{fieldSchema?.description}</TooltipContent>
-                  </Tooltip>
-                </TooltipProvider>
+  function isHandleConnected(handleId: string): boolean {
```
Comment (Abhi1992002) [inline autogpt_platform/frontend/src/app/(platform)/build/components/FlowEditor/nodes/OutputHandler.tsx:56]: 🟠 **Should Fix**: Missing `useMemo` for `connectedSourceHandles` and `outputTree`

Both are recomputed on every render. Since this component subscribes to all edges (as the other comment noted), it re-renders frequently — and each time it rebuilds the `Set` and calls `buildVisibleOutputTree` from scratch.

Even before fixing the edges selector, wrapping these in `useMemo` would reduce the cost of re-renders:

```ts
const connectedSourceHandles = useMemo(
  () =>
    new Set(
      edges
        .filter((edge) => edge.source === nodeId && edge.sourceHandle)
        .map((edge) => edge.sourceHandle as string),
    ),
  [edges, nodeId],
);

const outputTree = useMemo(
  () =>
    buildVisibleOutputTree({
      properties,
      isHandleConnected: (h) => connectedSourceHandles.has(h),
      isCollapsed: isOutputCollapsed,
    }),
  [properties, connectedSourceHandles, nodeOutputCollapsedStates, nodeId],
);
```

Without this, every edge change on the entire canvas triggers a full tree rebuild for every node — O(nodes × outputs) work per edge event.

File: autogpt_platform/frontend/src/app/(platform)/build/stores/nodeStore.ts [hunk 2]
```diff
@@ -126,6 +134,7 @@ export const useNodeStore = create<NodeStore>((set, get) => ({
   nodeCounter: 0,
   setNodeCounter: (nodeCounter) => set({ nodeCounter }),
   nodeAdvancedStates: {},
+  nodeOutputCollapsedStates: {},
```
Comment (Abhi1992002) [inline autogpt_platform/frontend/src/app/(platform)/build/stores/nodeStore.ts:137]: 🟡 **Nice to Have**: Collapsed state is not cleaned up when nodes are removed

When a node is deleted, its entry in `nodeOutputCollapsedStates[nodeId]` persists in the store. This is a minor memory leak — it won't cause bugs, but over a long editing session with many node adds/deletes, stale keys accumulate.

Consider adding cleanup in the node deletion logic, or adding a `clearOutputCollapsedState(nodeId)` action that gets called when nodes are removed.

File: autogpt_platform/frontend/src/app/(platform)/build/components/FlowEditor/nodes/output-visibility.test.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
+import { describe, expect, test } from "vitest";
```
Comment (Abhi1992002) [inline autogpt_platform/frontend/src/app/(platform)/build/components/FlowEditor/nodes/output-visibility.test.tsx:1]: 🔵 **Nit**: This test file uses `.tsx` but contains no JSX — `.test.ts` would be more accurate.

**Comment by ntindle** [comment]
closing for now due to unsigned cla :) feel free to reopen when signed :)
